flake.nix

```diff
@@ -0,0 +1,100 @@
+          self.callCabal2nix "categorifier-categories-integration-test"
+          ./integrations/categories/integration-test { };
+        "categorifier-concat-examples" =
+          self.callCabal2nix "categorifier-concat-examples"
+          ./integrations/concat/examples { };
+        "categorifier-concat-integration" =
+          self.callCabal2nix "categorifier-concat-integration"
+          ./integrations/concat/integration { };
+        "categorifier-concat-integration-test" =
+          self.callCabal2nix "categorifier-concat-integration-test"
+          ./integrations/concat/integration-test { };
+        "categorifier-concat-extensions-category" =
+          self.callCabal2nix "categorifier-concat-extensions-category"
+          ./integrations/concat-extensions/category { };
+        "categorifier-concat-extensions-integration" =
+          self.callCabal2nix "categorifier-concat-extensions-integration"
+          ./integrations/concat-extensions/integration { };
+        "categorifier-concat-extensions-integration-test" =
+          self.callCabal2nix "categorifier-concat-extensions-integration-test"
+          ./integrations/concat-extensions/integration-test { };
+        "categorifier-unconcat-category" =
+          self.callCabal2nix "categorifier-unconcat-category"
+          ./integrations/unconcat/category { };
+        "categorifier-unconcat-integration" =
+          self.callCabal2nix "categorifier-unconcat-integration"
+          ./integrations/unconcat/integration { };
+        "categorifier-unconcat-integration-test" =
+          self.callCabal2nix "categorifier-unconcat-integration-test"
+          ./integrations/unconcat/integration-test { };
+        "categorifier-vec-integration" =
+          self.callCabal2nix "categorifier-vec-integration"
+          ./integrations/vec/integration { };
+        "categorifier-vec-integration-test" =
+          self.callCabal2nix "categorifier-vec-integration-test"
+          ./integrations/vec/integration-test { };
+        "categorifier-plugin" =
+          self.callCabal2nix "categorifier-plugin" ./plugin { };
+        "categorifier-plugin-test" =
+          self.callCabal2nix "categorifier-plugin-test" ./plugin-test { };
+        "categorifier-th" = self.callCabal2nix "categorifier-th" ./th { };
+
+      };
+
+    in {
+      # see these issues and discussions:
+      # - https://github.com/NixOS/nixpkgs/issues/16394
+      # - https://github.com/NixOS/nixpkgs/issues/25887
+      # - https://github.com/NixOS/nixpkgs/issues/26561
+      # - https://discourse.nixos.org/t/nix-haskell-development-2020/6170
+      overlay = final: prev: {
+        haskellPackages = prev.haskellPackages.override (old: {
+          overrides = final.lib.composeExtensions (old.overrides or (_: _: { }))
+            haskellOverlay;
+        });
+      };
+
+      devShell.x86_64-linux = let
+        hsenv = pkgs.haskellPackages.ghcWithPackages
+          (p: [ p.cabal-install p.concat-examples ]);
```

Comment:
Why does this include `concat-examples`, but not `concat-classes` or any of our other dependencies?

One of the annoying things with `source-repository-package` is that we need to explicitly manage the transitive dependencies (like `concat-known`) that we don’t depend on directly. If that goes away, as it looks like it would, that is a big win.

Reply:
That's how the Nix system works. ;-) The dependency is transitively included as nix defines the closure of the corresponding nix package and the closure must come together. Of course, there is a little magic from ghcWithPackages which will define concatenated package set registered for GHC in that environment. It is all taken care automatically.

Reply:
It seems like here we should be using something like https://stackoverflow.com/questions/70649310/how-to-get-use-callcabal2nix-to-supply-package-list-to-ghcwithpackages instead of explicitly listing any dependencies of the cabal packages?

Reply:
Yes. The principled way is to traverse nix expressions of each package generated from callCabal2nix and collect `buildDepends` (Haskell cabal dependencies) and create a ghcWithPackages env with that. That's a common practice so we can do that without much trouble, i believe.